### PR TITLE
implement workbench: startup editor preference

### PR DESCRIPTION
**Pull request by vladarama**
#### What it does

Fixes: https://github.com/eclipse-theia/theia/issues/12754

This PR implements the workbench: startup editor preference. This allows the user to choose which editor is shown on startup when none are restored from the previous session.

Implemented preferences:
- `none`: no editor will be shown on startup.
- `welcomePage`: the default welcome page will be shown on startup.
- `readme`: it will open the folder's readme file in preview mode.
- `newUntitledFile`: opens an untitled file on startup.
- `welcomePageInEmptyWorkbench`: only opens the welcome page when opening
  empty workbench.

#### How to test

1.  Open Theia application.
2.  Go to `Settings` -> `workbench: startup editor` and select one of the preference to test.
3.  Close any opened editor and restart Theia.
4.  The selected editor should be shown on startup.
5.  Repeat 2-4 until all of the preferences are tested.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by vladarama** [comment]
Hey @msujew ,
Thanks for reviewing, I have addressed your comments and fixed the implementation errors. It should be a lot cleaner now 👍

**Review by msujew — APPROVED** [review]
Thanks, looks good to me. Each of the new preference values seem to work as expected 👍

**Comment by JonasHelming** [comment]
@vince-fugnitto Do you want to merge this?

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto Do you want to merge this?

Not yet, I'm still performing an offline review with @vladarama about the changes.

File: packages/core/src/browser/core-preferences.ts [hunk 1]
```diff
@@ -225,6 +225,19 @@ export const corePreferenceSchema: PreferenceSchema = {
...
+            enumDescriptions: [
+                nls.localizeByDefault('Start without an editor.'),
+                nls.localize('theia/getting-started/startup-editor/welcomePage', 'Open the Welcome page, with content to aid in getting started with {0} and extensions.', FrontendApplicationConfigProvider.get().applicationName),
+                nls.localizeByDefault('Open the README when opening a folder that contains one, fallback to \'welcomePage\' otherwise. Note: This is only observed as a global configuration, it will be ignored if set in a workspace or folder configuration.'),
+                nls.localizeByDefault('Open a new untitled text file (only applies when opening an empty window).'),
+                nls.localizeByDefault('Open the Welcome page when opening an empty workbench.'),
+            ],
+            default: 'welcomePage',
+            description: nls.localizeByDefault('Controls which editor is shown at startup, if none are restored from the previous session.')
+        },
```
Comment (vince-fugnitto) [inline packages/core/src/browser/core-preferences.ts:240]: Given that the implementation is completely present in `@theia/getting-started` I believe the preference should be moved there. With the current approach the preference would not be usable if the application does not include `@theia/getting-started`.

File: packages/getting-started/src/browser/getting-started-contribution.ts [hunk 2]
```diff
@@ -52,19 +75,51 @@ export class GettingStartedContribution extends AbstractViewContribution<Getting
...
+                        break;
+                    case 'welcomePageInEmptyWorkbench':
+                        if (!this.workspaceService.opened) {
+                            this.openView({ reveal: true, activate: true });
+                        }
+                        break;
+                    case 'newUntitledFile':
+                        const untitledUri = this.untitledResourceResolver.createUntitledURI('', await this.workingDirProvider.getUserWorkingDir());
+                        this.untitledResourceResolver.resolve(untitledUri);
+                        await open(this.openerService, untitledUri);
```
Comment (vince-fugnitto) [inline packages/getting-started/src/browser/getting-started-contribution.ts:94]: I believe we can simplify by re-using the `NEW_UNTITLED_TEXT_FILE` command:

```ts
this.commandRegistry.executeCommand(NEW_UNTITLED_TEXT_FILE.id)
```

File: packages/getting-started/src/browser/getting-started-preferences.ts [hunk 3]
```diff
@@ -22,21 +22,32 @@ import {
...
 
 export const GettingStartedPreferenceSchema: PreferenceSchema = {
     'type': 'object',
     properties: {
-        'welcome.alwaysShowWelcomePage': {
-            type: 'boolean',
-            description: nls.localizeByDefault('Show welcome page on startup'),
-            default: true
-        }
+        'welcome.startupEditor': {
```
Comment (vince-fugnitto) [inline packages/getting-started/src/browser/getting-started-preferences.ts:31]: We should use the same preference `id` present in VS Code in order to support their preference natively (reading from `.vscode/settings.json`):

```suggestion
        'workbench.startupEditor': {
```

File: packages/getting-started/src/browser/getting-started-preferences.ts [hunk 4]
```diff
@@ -22,21 +22,32 @@ import {
     PreferenceSchema,
     PreferenceContribution
 } from '@theia/core/lib/browser/preferences';
+import { FrontendApplicationConfigProvider } from '@theia/core/lib/browser/frontend-application-config-provider';
 import { nls } from '@theia/core/lib/common/nls';
 
 export const GettingStartedPreferenceSchema: PreferenceSchema = {
     'type': 'object',
     properties: {
-        'welcome.alwaysShowWelcomePage': {
```
Comment (vince-fugnitto) [inline packages/getting-started/src/browser/getting-started-preferences.ts:30]: Do you mind adding a breaking  changes note in our [changelog](https://github.com/eclipse-theia/theia/blob/master/CHANGELOG.md#v1410--) for v1.41.0 about the preference being removed in favor of `workbench.startupEditor`?

File: packages/getting-started/src/browser/getting-started-contribution.ts [hunk 5]
```diff
@@ -52,19 +78,49 @@ export class GettingStartedContribution extends AbstractViewContribution<Getting
...
-                    }
-                });
+                        break;
+                    case 'welcomePageInEmptyWorkbench':
+                        if (!this.workspaceService.opened) {
+                            this.openView({ reveal: true, activate: true });
+                        }
+                        break;
+                    case 'newUntitledFile':
+                        this.commandRegistry.executeCommand(CommonCommands.NEW_UNTITLED_FILE.id);
```
Comment (vince-fugnitto) [inline packages/getting-started/src/browser/getting-started-contribution.ts:95]: Close but I believe this is the wrong command, I mentioned `NEW_UNTITLED_TEXT_FILE` before :) It creates the file without the prompt for which type.

**Review by vince-fugnitto — APPROVED** [review]
LGTM :+1: I confirmed the following `workbench.startupEditor`:

| Preference | Use-Case | Result |
|:---|:---|:---|
| `welcomePage` | no editors present | the welcome page is opened on startup |
| `welcomePage` | no editors present (no workspace)| the welcome page is opened on startup |
| `welcomePage` | editors present | the welcome page is not opened on startup |
| `none` | all cases | nothing happens on startup besides restoring past views |
| `newUntitledFile` | no editors present | an untitled text file is opened on startup |
| `newUntitledFile` | no editors present (no workspace)| an untitled text file is opened on startup |
| `newUntitledFile` | editors present | no untitled file is opened on startup |
| `readme` | no editors present | the readme is opened on startup |
| `readme` | no editors present, no readme present in the workspace | the welcome page is opened as a fallback on startup |
| `readme` | no editors present (no workspace) | the welcome page is opened as a fallback on startup |
| `readme` | editors present | no readme is opened on startup |
| `welcomePageInEmptyWorkbench` | workspace present | no welcome page is opened on startup |
| `welcomePageInEmptyWorkbench` | no workspace present | the welcome page is opened on startup |

**Pull request merged**